### PR TITLE
feat(Forms): add `createMinimumAgeValidator` to Field.NationalIdentityNumber make a customizable `adultValidator`

### PR DESCRIPTION
**Pull request by tujoworker**
PR adds `createAgeValidator`.

Give the feature a test in this [CSB(outdated](https://codesandbox.io/p/sandbox/xenodochial-ardinghelli-ph6w2k?workspaceId=9eb653cb-4406-4b8f-bc96-714927cc7fcd) or in the [new CSB](https://codesandbox.io/p/sandbox/pensive-heisenberg-kcj484?workspaceId=9eb653cb-4406-4b8f-bc96-714927cc7fcd), or in the [pr preview](https://eufemia-git-feat-is-adult-validator-eufemia.vercel.app/uilib/extensions/forms/feature-fields/NationalIdentityNumber/demos/#extend-validation-with-adult-validator).

To dos:
- [x] Have a discussion on how we should export the validator(in terms of tree shaking).
- [x] Improve tests.
- [x] Add docs about it.

### CODE REVIEW

**Comment by langz** [comment]
I've added https://github.com/dnbexperience/eufemia/pull/4057/commits/0a91f30e85beac2400d82f930fa70e2410f31d85 which: 
- Makes a first stab at adult validator
- Adds tests for fnr, need to find for dnr)
- Adds example


TODO:
- Mock Date in tests, so that the tests is not dependent on the date today.

To be able to progress on this, given the example and test, we would have to solve https://github.com/dnbexperience/eufemia/issues/4073

File: packages/dnb-eufemia/src/extensions/forms/constants/locales/nb-NO.ts [hunk 1]
```diff
@@ -119,6 +119,7 @@ export default {
         'Ugyldig fødselsnummer. Skriv inn et gyldig fødselsnummer med 11 siffer.',
       errorFnr: 'Ugyldig fødselsnummer.',
       errorDnr: 'Ugyldig d-nummer.',
+      errorAdult: 'Ugyldig D number.',
```
Comment (langz) [inline packages/dnb-eufemia/src/extensions/forms/constants/locales/nb-NO.ts:122]: What should the error message be?

Comment (langz) [reply packages/dnb-eufemia/src/extensions/forms/constants/locales/nb-NO.ts:122]: image

File: packages/dnb-eufemia/src/extensions/forms/Field/NationalIdentityNumber/__tests__/NationalIdentityNumber.test.tsx [hunk 2]
```diff
@@ -6,6 +6,10 @@ import nbNO from '../../../constants/locales/nb-NO'
 
 const nb = nbNO['nb-NO']
 
+async function expectNever(callable: () => unknown): Promise<void> {
```
Comment (langz) [inline packages/dnb-eufemia/src/extensions/forms/Field/NationalIdentityNumber/__tests__/NationalIdentityNumber.test.tsx:9]: I’ll have to check if this is needed or not

File: packages/dnb-eufemia/src/extensions/forms/Field/NationalIdentityNumber/__tests__/NationalIdentityNumberAdultValidator.test.tsx [hunk 3]
```diff
@@ -0,0 +1,277 @@
+import React from 'react'
+import { render, waitFor, screen } from '@testing-library/react'
+import { Field, Validator } from '../../..'
+import nbNO from '../../../constants/locales/nb-NO'
+import userEvent from '@testing-library/user-event'
+
+const nb = nbNO['nb-NO']
+
+describe('Field.NationalIdentityNumber', () => {
```
Comment (langz) [inline packages/dnb-eufemia/src/extensions/forms/Field/NationalIdentityNumber/__tests__/NationalIdentityNumberAdultValidator.test.tsx:9]:
```suggestion
describe('Field.NationalIdentityNumber with adultValidator', () => {
```

**Comment by langz** [comment]
This looks good to me now @tujoworker 🚀 
Only thing left is perhaps the "discussion on how we should export the validator(in terms of tree shaking).", and if/how we should create a more generic validator like `isOfAge(age)` which could either replace `adultValidator` by `isOfAge(18)` or if we want both `isOfAge(age)`  and `adultValidator`.

**Comment by langz** [comment]
I think I might also extend the error messages provided in the validators, to also possibly return a specific message for when not providing 11 digits.

File: packages/dnb-design-system-portal/src/docs/uilib/extensions/forms/feature-fields/NationalIdentityNumber/properties.mdx [hunk 4]
```diff
@@ -21,11 +21,12 @@ import { NationalIdentityNumberProperties } from '@dnb/eufemia/src/extensions/fo
...
 `Field.NationalIdentityNumber` expose the following validators through its `validator` and `onBlurValidator` property:
 
-- `dnrValidator`: validates a d number.
+- `dnrValidator`: validates a D number.
 - `fnrValidator`: validates a national identity number (fødselsnummer).
 - `dnrAndFnrValidator`:
-  - validates the identification number as a d number when first digit is 4 or greater (because a d number has its first number increased by 4).
+  - validates the identification number as a D number when first digit is 4 or greater (because a D number has its first number increased by 4).
   - validates the identification number as a national identity number (fødselsnummer) when first digit is 3 or less.
+- `adultValidator`: validates if the identification number has a date of birth that is 18 years or older.
```
Comment (langz) [inline packages/dnb-design-system-portal/src/docs/uilib/extensions/forms/feature-fields/NationalIdentityNumber/properties.mdx:29]: Add some info about that it does not validate the corectnes of a identification number, only if the dage of birth used in the input is 18 years older, and that it’s highly likely that this should be used together with dnrAndFnrValidator to validate the validity of the identifikasjon number as well.

Comment (langz) [reply packages/dnb-design-system-portal/src/docs/uilib/extensions/forms/feature-fields/NationalIdentityNumber/properties.mdx:29]: I should also provide examples of only doing fnrValidator with AdiltValidator. Just to show what’s possible.

Comment (langz) [reply packages/dnb-design-system-portal/src/docs/uilib/extensions/forms/feature-fields/NationalIdentityNumber/properties.mdx:29]: Should also add some test for these combinations:
- fnrValidaror and adiultValidator
- - what heppens when providing dnr(i expect fnr error)
- dnrValidstor and adultValidator
- - what happens when providing fnr(i expect dnr error)

**Comment by tujoworker** [comment]
Is it ok for you when I push a change in for the validator? I think we are better off if we make a more flexible solution, like this:

```ts
const myAdultValidator = createAboveAgeValidator(18)
```

**Comment by langz** [comment]
> Is it ok for you when I push a change in for the validator? I think we are better off if we make a more flexible solution, like this:
> 
> ```ts
> const myAdultValidator = createAboveAgeValidator(18)
> ```

Sure, please go ahead :)
I just commited a small commit https://github.com/dnbexperience/eufemia/pull/4057/commits/8641d33ad3b11149202741ca44316594c41d643e
Hopefully it won't cause too many conflicts with yours 🙏

File: packages/dnb-eufemia/src/extensions/forms/Field/NationalIdentityNumber/NationalIdentityNumber.tsx [hunk 5]
```diff
@@ -54,6 +62,18 @@ function NationalIdentityNumber(props: Props) {
     [dnrValidator, fnrValidator]
   )
 
+  const adultValidator = useCallback(
+    (value: string) => {
+      // if (value !== undefined && value.length < 9) {
+      //   return Error(errorAdultPattern)
+      // }
```
Comment (langz) [inline packages/dnb-eufemia/src/extensions/forms/Field/NationalIdentityNumber/NationalIdentityNumber.tsx:69]: These lines can be removed

File: packages/dnb-eufemia/src/extensions/forms/Field/NationalIdentityNumber/__tests__/NationalIdentityNumberAdultValidator.test.tsx [hunk 6]
```diff
@@ -56,20 +60,18 @@ describe('Field.NationalIdentityNumber with adultValidator', () => {
     })
   })
 
-  it('should display error when validateInitially and value', async () => {
+  it.skip('should display error when value is invalid', async () => {
```
Comment (langz) [inline packages/dnb-eufemia/src/extensions/forms/Field/NationalIdentityNumber/__tests__/NationalIdentityNumberAdultValidator.test.tsx:63]: Todo

File: packages/dnb-eufemia/src/extensions/forms/Field/NationalIdentityNumber/__tests__/NationalIdentityNumberAdultValidator.test.tsx [hunk 7]
```diff
@@ -0,0 +1,665 @@
...
+      )
+    })
+  })
+
+  it.skip('should display error when value is invalid', async () => {
+    render(
+      <Field.NationalIdentityNumber
+        onBlurValidator={myAdultValidator}
+        validateInitially
+        value="123" // outputs the age of 105 years, which is valid
```
Comment (langz) [inline packages/dnb-eufemia/src/extensions/forms/Field/NationalIdentityNumber/__tests__/NationalIdentityNumberAdultValidator.test.tsx:68]: 123 shouldn’t output 105 years?
All info we have based on 123 is that person is born 12th in a month starting with 3, which doesn’t exist?🤔

File: packages/dnb-eufemia/src/extensions/forms/Field/NationalIdentityNumber/NationalIdentityNumber.tsx [hunk 8]
```diff
@@ -106,5 +114,57 @@ function NationalIdentityNumber(props: Props) {
...
+  const day = isDnr
+    ? Number.parseInt(value.substring(0, 2)) - 40
+    : Number.parseInt(value.substring(0, 2))
+
+  return new Date(Number.parseInt(year), month - 1, day)
+}
+
+export function createAboveAgeValidator(age: number) {
+  return (value: string) => {
+    if (typeof value !== 'string' || value.length < 9) {
```
Comment (langz) [inline packages/dnb-eufemia/src/extensions/forms/Field/NationalIdentityNumber/NationalIdentityNumber.tsx:155]:
```suggestion
    if (typeof value !== 'string' || value.length < 7) {
```

File: packages/dnb-eufemia/src/extensions/forms/Field/NationalIdentityNumber/__tests__/NationalIdentityNumberAdultValidator.test.tsx (hunk 7 above)
Comment (tujoworker) [reply packages/dnb-eufemia/src/extensions/forms/Field/NationalIdentityNumber/__tests__/NationalIdentityNumberAdultValidator.test.tsx]: The logic needed to be adjusted so the age check is skipped. Now the error message still hits, when there is no possibility of calculate a correct age.

File: packages/dnb-eufemia/src/extensions/forms/Field/NationalIdentityNumber/__tests__/NationalIdentityNumberAdultValidator.test.tsx (hunk 6 above)
Comment (tujoworker) [reply packages/dnb-eufemia/src/extensions/forms/Field/NationalIdentityNumber/__tests__/NationalIdentityNumberAdultValidator.test.tsx]: Done 👍

**Comment by tujoworker** [comment]
I think, having the name "Above" can be misleading. Any better suggestions are welcome. Else we just drop the word above in the `createAgeValidator`.

**Comment by langz** [comment]
I guess we should update docs for `adultValidator` as it’s not a validator named that anymore?

**Comment by tujoworker** [comment]
Yeah, this could be adjusted. Can you take a look at it?

File: packages/dnb-eufemia/src/extensions/forms/constants/locales/en-GB.ts [hunk 9]
```diff
@@ -120,6 +120,7 @@ export default {
       errorRequired: 'You must enter a national identity number.',
       errorFnr: 'Invalid national identity number.',
       errorDnr: 'Invalid D number.',
+      errorBelowAge: 'Must be at least {age} years of age.',
```
Comment (langz) [inline packages/dnb-eufemia/src/extensions/forms/constants/locales/en-GB.ts:123]:
```suggestion
      ageValidatorError: 'Must be at least {age} years of age.',
```
Maybe this is more in line of how we’ve named the other translations?

Comment (langz) [reply packages/dnb-eufemia/src/extensions/forms/constants/locales/en-GB.ts:123]: I'll make the change then.

**Comment by langz** [comment]
To me, it looks good, 
I've validated that the [old CSB(before making the reusable function createMinimumAgeValidator](https://codesandbox.io/p/sandbox/xenodochial-ardinghelli-ph6w2k?workspaceId=9eb653cb-4406-4b8f-bc96-714927cc7fcd) works the same as in our [new CSB(when using the createMinimumAgeValidator)](https://codesandbox.io/p/sandbox/pensive-heisenberg-kcj484?workspaceId=9eb653cb-4406-4b8f-bc96-714927cc7fcd),

File: packages/dnb-design-system-portal/src/docs/uilib/extensions/forms/feature-fields/NationalIdentityNumber/properties.mdx [hunk 10]
```diff
@@ -21,12 +21,25 @@ import { NationalIdentityNumberProperties } from '@dnb/eufemia/src/extensions/fo
...
-- `dnrValidator`: validates a d number.
+- `dnrValidator`: validates a D number.
 - `fnrValidator`: validates a national identity number (fødselsnummer).
 - `dnrAndFnrValidator`:
-  - validates the identification number as a d number when first digit is 4 or greater (because a d number has its first number increased by 4).
+  - validates the identification number as a D number when first digit is 4 or greater (because a D number has its first number increased by 4).
   - validates the identification number as a national identity number (fødselsnummer) when first digit is 3 or less.
 
+You can create your own age validator by using the `createMinimumAgeValidator` function. It takes an age as a parameter and returns a validator function. The validator function takes a value and returns an error message if the value is not above the given age.
+It validates if the identification number has a date of birth that is 18 years or older. It uses only the 7 first digits of an identification number to validate, first 6 digits representing the birth of date, and the next digit represents the century.
```
Comment (langz) [inline packages/dnb-design-system-portal/src/docs/uilib/extensions/forms/feature-fields/NationalIdentityNumber/properties.mdx:31]:
```suggestion
It validates if the identification number has a date of birth that is 18 years or older. It uses only the 7 first digits of the identification number to validate. The first 6 digits representing the birth of date, and the next digit represents the century.
```

File: packages/dnb-design-system-portal/src/docs/uilib/extensions/forms/feature-fields/NationalIdentityNumber/properties.mdx [hunk 11]
```diff
@@ -21,12 +21,25 @@ import { NationalIdentityNumberProperties } from '@dnb/eufemia/src/extensions/fo
...
+As it only use the 7 first digits, it does not validate the identification number, therefore it's quite common to use this validator together with one of the validators above (`dnrValidator`, `fnrValidator` or `dnrAndFnrValidator`) to validate the identification number as well.
+
+You need to import the `createMinimumAgeValidator` function from the `Field.NationalIdentityNumber` component:
+
+```tsx
+import { createMinimumAgeValidator } from '@dnb/eufemia/extensions/forms/Field/NationalIdentityNumber'
+
+// Create a validator that validates if the value is above 18 years old
+const above18Validator = createMinimumAgeValidator(18)
```
Comment (langz) [inline packages/dnb-design-system-portal/src/docs/uilib/extensions/forms/feature-fields/NationalIdentityNumber/properties.mdx:40]:
```suggestion
const above18YearsValidator = createMinimumAgeValidator(18)
```

Comment (tujoworker) [reply packages/dnb-design-system-portal/src/docs/uilib/extensions/forms/feature-fields/NationalIdentityNumber/properties.mdx]: The word above is a bit misleading. It's inclusive and above.

**Pull request merged**